storage/remote/write_handler.go

```diff
@@ -256,87 +269,178 @@ func (h *writeHandler) write(ctx context.Context, req *prompb.WriteRequest) (err
+// Once we have 5xx type of error, we immediately stop and rollback all appends.
+func (h *writeHandler) writeV2(ctx context.Context, req *writev2.Request) (_ responseStats, _ error, httpCode int) {
+	app := &timeLimitAppender{
 		Appender: h.appendable.Appender(ctx),
 		maxTime:  timestamp.FromTime(time.Now().Add(maxAheadTime)),
 	}
 
-	defer func() {
-		if err != nil {
-			_ = timeLimitApp.Rollback()
-			return
+	rs := responseStats{}
+	if err, httpCode := h.appendV2(app, req, &rs); err != nil {
+		if httpCode/5 == 100 {
+			// On 5xx, we always rollback, because we expect
+			// sender to retry and TSDB is not idempotent.
+			if rerr := app.Rollback(); rerr != nil {
+				level.Error(h.logger).Log("msg", "writev2 rollback failed on retry-able error", "err", rerr)
+			}
+			return responseStats{}, err, httpCode
 		}
-		err = timeLimitApp.Commit()
-	}()
 
-	b := labels.NewScratchBuilder(0)
+		// Non-retriable (e.g. bad request error case). Can be partially written.
+		commitErr := app.Commit()
+		if commitErr != nil {
+			// Bad requests does not matter as we have internal error (retryable).
+			return responseStats{}, commitErr, http.StatusInternalServerError
+		}
+
+		// Bad request error happened, but rest of data (if any) was written.
+		return rs, err, httpCode
+	}
+
+	// All good just commit.
+	if err := app.Commit(); err != nil {
+		return responseStats{}, err, http.StatusInternalServerError
+	}
+	return rs, nil, 0
+}
+
+func (h *writeHandler) appendV2(app storage.Appender, req *writev2.Request, rs *responseStats) (err error, httpCode int) {
+	var (
+		badRequestErrs                                   []error
+		outOfOrderExemplarErrs, samplesWithInvalidLabels int
+
+		b = labels.NewScratchBuilder(0)
+	)
 	for _, ts := range req.Timeseries {
 		ls := ts.ToLabels(&b, req.Symbols)
+		// Validate series labels early.
+		// NOTE(bwplotka): While spec allows UTF-8, Prometheus Receiver may impose
+		// specific limits and follow https://prometheus.io/docs/specs/remote_write_spec_2_0/#invalid-samples case.
+		if !ls.IsValid() {
+			badRequestErrs = append(badRequestErrs, fmt.Errorf("invalid metric name or labels, got %v", ls.String()))
+			samplesWithInvalidLabels += len(ts.Samples) + len(ts.Histograms)
+			continue
+		}
 
-		err := h.appendSamplesV2(timeLimitApp, ts.Samples, ls)
-		if err != nil {
-			return err
+		var ref storage.SeriesRef
+
+		// Samples.
+		for _, s := range ts.Samples {
+			ref, err = app.Append(ref, ls, s.GetTimestamp(), s.GetValue())
+			if err == nil {
+				rs.samples++
+				continue
+			}
+
+			// Handle append error.
+			if errors.Is(err, storage.ErrOutOfOrderSample) ||
+				errors.Is(err, storage.ErrOutOfBounds) ||
+				errors.Is(err, storage.ErrDuplicateSampleForTimestamp) ||
+				errors.Is(err, storage.ErrTooOldSample) {
+				// TODO(bwplotka): Not too spammy log?
+				level.Error(h.logger).Log("msg", "Out of order sample from remote write", "err", err.Error(), "series", ls.String(), "timestamp", s.Timestamp)
```

Comment:
potentially, can we use the dedupe logger here?

Reply:
What's dedupe logger? 🤔

Reply:
Hm, I see https://github.com/prometheus/prometheus/blob/c954cd9d1d4e3530be2939d39d8633c38b70913f/util/logging/dedupe.go#L54

Reply:
since we log timestamp, series etc, this will not help I think.. I would say let's fix in further PRs if this becomes a problem somewhere.